ramips: add support for RAISECOM MSG1500 X.00

ylk951207 opened the pull request:
RAISECOM MSG1500 X.00 is a 2.4/5 GHz band 11ac (Wi-Fi 5) router.
Apart from the general model, there are two ISP customized models:
China Mobile and China Telecom.

Specifications:

- SoC: Mediatek MT7621AT
- RAM: 256MiB DDR3
- Flash: 128MiB NAND
- Ethernet: 5 * 10/100/1000Mbps: 4 * LAN + 1 * WAN
  - Switch: MediaTek MT7530 (SoC)
- WLAN: 1 * MT7615DN Dual-Band 2.4GHz 2T2R (400Mbps) 5GHz 2T2R (867Mbps)
- USB: 1 * USB 2.0 port
- Button: 1 * RESET button, 1 * WPS button, 1 * WIFI button
- LED: blue color: POWER, WAN, WPS, 2.4G, 5G, LAN1, LAN2, LAN3, LAN4, USB
- UART: 1 * serial port header (4-pin)
- Power: DC 12V, 1A
  - Switch: 1 * POWER switch

MAC addresses as verified by vendor firmware:

use   address             source
LAN   C8:XX:XX:3A:XX:E7   Config   "protest_lan_mac"  ascii  (label)
WAN   C8:XX:XX:3A:XX:EA   Config   "protest_wan_mac"  ascii
5G    C8:XX:XX:3A:XX:E8   Factory  "0x4"              hex
2.4G  CA:XX:XX:4A:XX:E8   [not on flash]

The increment of the 4th byte for the 2.4g address appears to vary.
Reported cases:

       5g                 2.4g         increment
 C8:XX:XX:90:XX:C3  CA:XX:XX:C0:XX:C3  0x30
 C8:XX:XX:3A:XX:08  CA:XX:XX:4A:XX:08  0x10
 C8:XX:XX:3A:XX:E8  CA:XX:XX:4A:XX:E8  0x10

Since increment is inconsistent and there is no obvious pattern
in swapping bytes, and the 2.4g address has local bit set anyway,
it seems safer to use the LAN address with flipped byte here in
order to prevent collisions between OpenWrt devices and OEM devices
for this interface. This way we at least use an address as base
that is definitely owned by the device at hand.

Notes:

1. The vendor firmware allows you to connect to the router by telnet.
   (known version 1.0.0 can open telnet.)
   There is no official binary firmware available.
   Backup the important partitions data:
   "Bootloader", "Config", "Factory", and "firmware".
   Note that with the vendor firmware the memory is detected only 128MiB
   and the last 512KiB in NAND flash is not used.

2. The POWER LED is default on after press POWER switch.
   The WAN and LAN1 - 4 LEDs are wired to ethernet switch.
   The WPS LED is controlled by MT7615DN's GPIO.
   Currently there is no proper way to configure it.

3. At the time of adding support the wireless config needs to be set up 
   by editing the wireless config file:

 * Setting the country code is mandatory, otherwise the router loses
   connectivity at the next reboot. This is mandatory and can be done
   from luci. After setting the country code the router boots correctly.
   A reset with the reset button will fix the issue and the user has to
   reconfigure.

 * This is minor since the 5g interface does not come up online although
   it is not set as disabled. 2 options here:

   1- Either run the "wifi" command. Can be added from LuCI in system -
      startup - local startup and just add wifi above "exit 0".

   2- Or add the serialize option in the wireless config file as shown
      below. This one would work and bring both interfaces automatically
      at every boot:

      config wifi-device 'radio0'
          option serialize '1'

      config wifi-device 'radio1'
          option serialize '1'

Flash instructions using initramfs image:

1. Press POWER switch to power down if the router is running.

2. Connect PC to one of LAN ports, and set
   static IP address to "10.10.10.2", netmask to "255.255.255.0",
   and gateway to "10.10.10.1" manually on the PC.

3. Push and hold the WIFI button, and then power up the router.
   After about 10s (or you can call the recovery page, see "4" below)
   you can release the WIFI button.
   There is no clear indication when the router
   is entering or has entered into "RAISECOM Router Recovery Mode".

4. Call the recovery page for the router at "http://10.10.10.1".
   Keep an eye on the "WARNING!! tip" of the recovery page.
   Click "Choose File" to select initramfs image, then click "Upload".

5. If image is uploaded successfully, you will see the page display
   "Device is upgrading the firmware... %".
   Keep an eye on the "WARNING!! tip" of the recovery page.
   When the page display "Upgrade Successfully",
   you can set IP address as "automatically obtain".

6. After the rebooting (PC should automatically obtain an IP address),
   open the SSH connection, then download the sysupgrade image
   to the router and perform sysupgrade with it.

Flash back to vendor firmware:

 See "Flash instructions 1 - 5" above.
 The only difference is that in step 4
 you should select the vendor firmware which you backup.

Signed-off-by: Liangkuan Yang <ylk951207@gmail.com>

adschm commented on target/linux/ramips/dts/mt7621_raisecom_msg1500-x.00.dts:
```diff
@@ -0,0 +1,146 @@
+#include <dt-bindings/input/input.h>
+
+/ {
+	compatible = "raisecom,msg1500-x.00", "mediatek,mt7621-soc";
```
I'd prefer using minus instead of dot here as well, since using the dot should stay an exception:
"raisecom,msg1500-x-00"

ylk951207 replied:
Thanks so much for your advice.

adschm commented on target/linux/ramips/dts/mt7621_raisecom_msg1500-x.00.dts:
```diff
@@ -0,0 +1,146 @@
+		compatible = "gpio-leds";
+
+		2.4g {
+			label = "blue:2.4g";
```
Most of the other devices use other names like wlan24g, wifi2g, etc. I'd prefer to not introduce a "new name" here, but choose one of the abundant ones ...

ylk951207 replied:
Thanks for pointing out that.

adschm commented on target/linux/ramips/mt7621/base-files/etc/board.d/01_leds:
```diff
@@ -78,6 +78,10 @@ netgear,r6800)
 	;;
+raisecom,msg1500-x.00)
+	ucidef_set_led_netdev "2.4g" "2.4G" "blue:2.4g" "wlan0"
+	ucidef_set_led_netdev "5g" "5G" "blue:5g" "wlan1"
```
Please use DT triggers instead.

ylk951207 replied:
OK. And I add "link" to keep the same logic with the vendor firmware.

adschm commented:
Is the dual-band chip simultaneous, so would we need to set the second MAC address?
If yes, consider trying https://github.com/openwrt/openwrt/blob/master/target/linux/ramips/mt7621/base-files/etc/hotplug.d/ieee80211/10_fix_wifi_mac

ylk951207 commented:
> Is the dual-band chip simultaneous, so would we need to set the second MAC address?
> If yes, consider trying https://github.com/openwrt/openwrt/blob/master/target/linux/ramips/mt7621/base-files/etc/hotplug.d/ieee80211/10_fix_wifi_mac

Yes, but I don't know how to change "C8:50:E9:3A" to "CA:50:E9:4A". It seems that "C8" to "CA" can use "macaddr_setbit_la". Sorry for that I'm a beginner.

ylk951207 commented:
> Is the dual-band chip simultaneous, so would we need to set the second MAC address?
> If yes, consider trying https://github.com/openwrt/openwrt/blob/master/target/linux/ramips/mt7621/base-files/etc/hotplug.d/ieee80211/10_fix_wifi_mac

Yes, but I don't know how to change "C8:50:E9:3A" to "CA:50:E9:4A". It seems that "C8" to "CA" can use "macaddr_setbit_la". Sorry for that I am a noob.

adschm commented on target/linux/ramips/mt7621/base-files/etc/board.d/01_leds:
```diff
@@ -78,6 +78,10 @@ netgear,r6800)
 	;;
+raisecom,msg1500-x-00)
+	ucidef_set_led_netdev "wlan2g" "Wi-Fi 2.4G" "blue:wlan2g" "wlan0" "link"
+	ucidef_set_led_netdev "wlan5g" "Wi-Fi 5G" "blue:wlan5g" "wlan1" "link"
```
_DTS_ trigger, e.g. https://github.com/openwrt/openwrt/blob/master/target/linux/ramips/dts/mt7621_tplink_rexx0-v1.dtsi#L66

This should replace the entries in 01_leds entirely, so they can be dropped.

ylk951207 replied:
Thanks for your advice. "phyxradio" is appropriate after testing.

adschm commented:
> Yes, but I don't know how to change "C8:50:E9:3A" to "CA:50:E9:4A". It seems that "C8" to "CA" can use "macaddr_setbit_la". Sorry for that I am a noob.

Well, technically you would use `macaddr_setbit_la $(macaddr_add $baseaddr 0x100000)`

Initially, I wondered whether they really calculate it and if it's not coming from somewhere else. But since it's a local address, it's probably actually calculated. And since 3 becomes 4, it's also not a bit flipped, but actually an increment.

ylk951207 commented:
> > Yes, but I don't know how to change "C8:50:E9:3A" to "CA:50:E9:4A". It seems that "C8" to "CA" can use "macaddr_setbit_la". Sorry for that I am a noob.
> 
> Well, technically you would use `macaddr_setbit_la $(macaddr_add $baseaddr 0x100000)`
> 

Thanks for your advice.
> Initially, I wondered whether they really calculate it and if it's not coming from somewhere else. But since it's a local address, it's probably actually calculated. And since 3 becomes 4, it's also not a bit flipped, but actually an increment.

[https://github.com/hanwckf/rt-n56u/blob/master/trunk/proprietary/rt_wifi/rtpci/5.0.5.1/mt7615/hw_ctrl/cmm_chip_mt.c#L731](url)
[https://github.com/hanwckf/rt-n56u/blob/master/trunk/proprietary/rt_wifi/rtpci/7.3.0.1/mt7915/hw_ctrl/cmm_chip_mt.c#L375
](url)I found that 2.4GHz macaddr is calculated by the function "MtAsicSetMbssWdevIfAddrGen2".
With the vendor firmware "MacMask" is "0x0f" and idx is "4".
Sorry for that I don't know how to write the code.

ylk951207 commented:
We can configure mt7615d dbdc mode via LuCI after OpenWrt commit e1d57d4 and LuCI commit fb82e03, so I update the commit message. Thanks for all of you.

ylk951207 committed:
ramips: add support for RAISECOM MSG1500 X.00

RAISECOM MSG1500 X.00 is a 2.4/5 GHz band 11ac (Wi-Fi 5) router.
Apart from the general model, there are two ISP customized models:
China Mobile and China Telecom.

Specifications:

- SoC: Mediatek MT7621AT
- RAM: 256MiB DDR3
- Flash: 128MiB NAND
- Ethernet: 5 * 10/100/1000Mbps: 4 * LAN + 1 * WAN
  - Switch: MediaTek MT7530 (SoC)
- WLAN: 1 * MT7615DN Dual-Band 2.4GHz 2T2R (400Mbps) 5GHz 2T2R (867Mbps)
- USB: 1 * USB 2.0 port
- Button: 1 * RESET button, 1 * WPS button, 1 * WIFI button
- LED: blue color: POWER, WAN, WPS, 2.4G, 5G, LAN1, LAN2, LAN3, LAN4, USB
- UART: 1 * serial port header (4-pin)
- Power: DC 12V, 1A
  - Switch: 1 * POWER switch

MAC addresses as verified by vendor firmware:

use   address             source
LAN   C8:XX:XX:3A:XX:E7   Config   "protest_lan_mac"  ascii  (label)
WAN   C8:XX:XX:3A:XX:EA   Config   "protest_wan_mac"  ascii
5G    C8:XX:XX:3A:XX:E8   Factory  "0x4"              hex
2.4G  CA:XX:XX:4A:XX:E8   [not on flash]

The increment of the 4th byte for the 2.4g address appears to vary.
Reported cases:

       5g                 2.4g         increment
 C8:XX:XX:90:XX:C3  CA:XX:XX:C0:XX:C3  0x30
 C8:XX:XX:3A:XX:08  CA:XX:XX:4A:XX:08  0x10
 C8:XX:XX:3A:XX:E8  CA:XX:XX:4A:XX:E8  0x10

Since increment is inconsistent and there is no obvious pattern
in swapping bytes, and the 2.4g address has local bit set anyway,
it seems safer to use the LAN address with flipped byte here in
order to prevent collisions between OpenWrt devices and OEM devices
for this interface. This way we at least use an address as base
that is definitely owned by the device at hand.

Notes:

1. The vendor firmware allows you to connect to the router by telnet.
   (known version 1.0.0 can open telnet.)
   There is no official binary firmware available.
   Backup the important partitions data:
   "Bootloader", "Config", "Factory", and "firmware".
   Note that with the vendor firmware the memory is detected only 128MiB
   and the last 512KiB in NAND flash is not used.

2. The POWER LED is default on after press POWER switch.
   The WAN and LAN1 - 4 LEDs are wired to ethernet switch.
   The WPS LED is controlled by MT7615DN's GPIO.
   Currently there is no proper way to configure it.

3. At the time of adding support the wireless config needs to be set up
   by editing the wireless config file:

 * Setting the country code is mandatory, otherwise the router loses
   connectivity at the next reboot. This is mandatory and can be done
   from luci. After setting the country code the router boots correctly.
   A reset with the reset button will fix the issue and the user has to
   reconfigure.

 * This is minor since the 5g interface does not come up online although
   it is not set as disabled. 2 options here:

   1- Either run the "wifi" command. Can be added from LuCI in system -
      startup - local startup and just add wifi above "exit 0".

   2- Or add the serialize option in the wireless config file as shown
      below. This one would work and bring both interfaces automatically
      at every boot:

      config wifi-device 'radio0'
          option serialize '1'

      config wifi-device 'radio1'
          option serialize '1'

Flash instructions using initramfs image:

1. Press POWER switch to power down if the router is running.

2. Connect PC to one of LAN ports, and set
   static IP address to "10.10.10.2", netmask to "255.255.255.0",
   and gateway to "10.10.10.1" manually on the PC.

3. Push and hold the WIFI button, and then power up the router.
   After about 10s (or you can call the recovery page, see "4" below)
   you can release the WIFI button.
   There is no clear indication when the router
   is entering or has entered into "RAISECOM Router Recovery Mode".

4. Call the recovery page for the router at "http://10.10.10.1".
   Keep an eye on the "WARNING!! tip" of the recovery page.
   Click "Choose File" to select initramfs image, then click "Upload".

5. If image is uploaded successfully, you will see the page display
   "Device is upgrading the firmware... %".
   Keep an eye on the "WARNING!! tip" of the recovery page.
   When the page display "Upgrade Successfully",
   you can set IP address as "automatically obtain".

6. After the rebooting (PC should automatically obtain an IP address),
   open the SSH connection, then download the sysupgrade image
   to the router and perform sysupgrade with it.

Flash back to vendor firmware:

 See "Flash instructions 1 - 5" above.
 The only difference is that in step 4
 you should select the vendor firmware which you backup.

Signed-off-by: Liangkuan Yang <ylk951207@gmail.com>

ylk951207 commented:
@adschm 
I found another 2.4GHz macaddr of the device.

5GHz: C8:XX:XX:90:XX:C3
2.4GHz: CA:XX:XX:C0:XX:C3
increment: 0x30

It seems that there is no regularity of the 2.4GHz macaddr. I'm sorry for my previous opinion. The case of 2.4GHz macaddr looks like dir-853-a3, so refering to it I modify the commit.

hauke commented:
Thank you for the patch, I applied it to master in 766733e17226221a310a30c91927a79f07a456e5.
